picard/util/pipe.py

```diff
@@ -0,0 +1,232 @@
+        finally:
+            win32file.CloseHandle(pipe)
+
+        return True
+
+    def __unix_sender(self, message: str) -> bool:
+        with open(self.path, 'a') as fifo:
+            fifo.write(message)
+        return True
+
+    def send_to_pipe(self, message: str, timeout_secs: float = 1.5) -> bool:
+        __pool = concurrent.futures.ThreadPoolExecutor()
+        sender = None
+        if self.__is_win:
+            sender = __pool.submit(self.__win_sender, message)
+        else:
+            sender = __pool.submit(self.__unix_sender, message)
+
+        try:
+            if sender.result(timeout=timeout_secs):
+                return True
+        except concurrent.futures._base.TimeoutError:
+            # hacky way to kill the sender
+            self.read_from_pipe()
+
+        return False
+
+    def read_from_pipe(self, timeout_secs: float = 1.5) -> str:
+        if self.__is_win:
+            return str(self.__read_from_win_pipe(timeout_secs))
+        else:
+            return self.__read_from_unix_pipe(timeout_secs)
+
+    def __win_reader(self) -> str:
+        response = ""  # type: ignore
+
+        try:
+            pipe = win32file.CreateFile(
+                self.path,
+                win32file.GENERIC_READ | win32file.GENERIC_WRITE,
+                self.__SHARE_MODE,
+                None,
+                win32file.OPEN_EXISTING,
+                self.__FLAGS_AND_ATTRIBUTES,
+                None
+            )
+            while not response:
+                response = win32file.ReadFile(pipe, 64 * 1024)
+
+        except WinApiError as err:
+            if err.args[0] == 2:
+                raise FileNotFoundError(Pipe.NOT_FOUND_MESSAGE)
+            elif err.args[0] == 109:
+                raise FileNotFoundError("Pipe is broken")
+            else:
+                raise FileNotFoundError(f"{err.args[0]}; {err.args[1]}; {err.args[2]}")
+
+        if response:
+            if response[0] == 0:
+                return response[1].decode("utf-8")  # type: ignore
+            else:
+                raise ValueError(f"INVALID RESPONSE: {response[1].decode('utf-8')}")  # type: ignore
+        else:
+            return Pipe.NO_RESPONSE_MESSAGE
+
+    def __read_from_win_pipe(self, timeout_secs: float) -> str:
+        __pool = concurrent.futures.ThreadPoolExecutor()
+        reader = __pool.submit(self.__win_reader)
+
+        try:
+            if reader.result(timeout=timeout_secs):
+                res: str = reader.result()
+                if res != self.MESSAGE_TO_IGNORE:
+                    return res
+        except concurrent.futures._base.TimeoutError:
+            # hacky way to kill the file-opening loop
+            self.send_to_pipe(self.MESSAGE_TO_IGNORE)
+
+        return Pipe.NO_RESPONSE_MESSAGE
+
+    def __unix_reader(self) -> str:
+        response: str = ""
+        while not response:
+            try:
+                fifo = open(self.path, 'r')
```

Comment:
Is there a reason to not use `with open(self.path, 'r') as fifo:` here ?

Reply:
None too, fixed